src/main/java/net/imagej/ops/segment/hough/HoughCircleDetectorLocalExtremaOp.java

```diff
@@ -0,0 +1,216 @@
+
+				for ( final T t : neighborhood )
+					if ( t.compareTo( c ) > 0 )
+						return null;
+
+				final double val = c.getRealDouble();
+				final double radius = minRadius + ( center.getDoublePosition( numDimensions - 1 ) ) * stepRadius;
+				final double ls = 2. * Math.PI * radius / val;
+				return new Circle( center, radius, ls );
+			}
+		};
+		final ArrayList< Circle > peaks = LocalExtrema.findLocalExtrema( input, maximumCheck, threadService.getExecutorService() );
+
+		if ( isCanceled() )
+			return Collections.emptyList();
+
+		/*
+		 * Non-maxima suppression.
+		 * 
+		 * Rule: when one circle has a center inside one another, we discard the
+		 * one with the highest sensitivity.
+		 */
+
+		// Sort by ascending sensitivity.
+		Collections.sort( peaks );
+
+		final List< Circle > retained = new ArrayList<>();
+		NEXT_CIRCLE: for ( final Circle tested : peaks )
+		{
+			for ( final Circle kept : retained )
+			{
+				if ( kept.contains( tested ) )
+					continue NEXT_CIRCLE;
+			}
+
+			// Was not found in any circle, so we keep it.
+			retained.add( tested );
+		}
+
+		if ( isCanceled() )
+			return Collections.emptyList();
+
+		/*
+		 * Refine local extrema.
+		 */
+
+		final SubpixelLocalization< Circle, T > spl = new SubpixelLocalization<>( numDimensions );
+		spl.setAllowMaximaTolerance( true );
+		spl.setMaxNumMoves( 10 );
+		final ArrayList< RefinedPeak< Circle > > refined = spl.process( retained, input, input );
+
+		if ( isCanceled() )
+			return Collections.emptyList();
+
+		/*
+		 * Create circles.
+		 */
+
+		final ArrayList< HoughCircle > circles = new ArrayList<>( refined.size() );
+		for ( final RefinedPeak< Circle > peak : refined )
+		{
+			final double radius = minRadius + ( peak.getDoublePosition( numDimensions - 1 ) ) * stepRadius;
+			final double ls = 2. * Math.PI * radius / peak.getValue();
+			if ( ls < 0 || ls > sensitivity )
+				continue;
+			final RealPoint center = new RealPoint( numDimensions - 1 );
+			for ( int d = 0; d < numDimensions - 1; d++ )
+				center.setPosition( peak.getDoublePosition( d ), d );
+
+			circles.add( new HoughCircle( center, radius, ls ) );
+		}
+
+		Collections.sort( circles );
+		return circles;
+	}
+
+	// -- Cancelable methods --
+
+	/** Reason for cancelation, or null if not canceled. */
+	private String cancelReason;
+
+	@Override
+	public boolean isCanceled()
+	{
+		return cancelReason != null;
+	}
+
+	/** Cancels the command execution, with the given reason for doing so. */
+	@Override
+	public void cancel( final String reason )
+	{
+		cancelReason = reason == null ? "" : reason;
+	}
+
+	@Override
+	public String getCancelReason()
+	{
+		return cancelReason;
+	}
+
+	/*
+	 * INNER CLASSES
+	 */
+
+	/**
+	 * A circle class defined on integer position with a sensitivity value.
+	 * 
+	 * @author Jean-Yves Tinevez
+	 *
+	 */
+	private final class Circle extends Point implements Comparable< Circle >
+	{
+		private final double radius;
+
+		private final double sensitivity;
```

Comment:
Can you also explain to me what sensitivity is measuring?

Reply:
Hough circle detection is a detection algorithm. Many of them returns a quality measure for what they detect (spot, circle, line,...) that reports how likely it is that it is not a spurious detection. The greater the quality, the more likely that it is a real one.

In this case I chose to use `sensitivity` which varies as the inverse of this quality. The sensitivity of a circle appears as the minimal value the _sensitivity settings_ must be set to incorporate it in the results.

In the UI, the user would change a sensitivity settings, and all the circles with a sensitivity _higher_ than this threshold would be discarded.

Reply:
Aha! Now I understand.